File: joongang/4949.py
from sys import stdin
while 1:
    t = stdin.readline().rstrip()
    if t == '.':
        break
    stack = []
    chk = 0
    for i in t:
        if i == '(' or i == '[':
            stack.append(i)
        if i == ')':
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                print('no')
                chk = 1
                break
        if i == ']':
            if stack and stack[-1] == '[':
                stack.pop()
            else:
                print('no')
                chk = 1
                break
    if not chk:
        if not stack:
            print('yes')
        else:
            print('no')

File: civilian38/4949.java
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main{
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String line = br.readLine();
        while (line != null) {
            if(line.charAt(0) == '.')
                break;
            boolean checker = true;
            int i = 0;
            Stack<Character> parentheses = new Stack<>();
            while (line.charAt(i) != '.'){
                if (line.charAt(i) == '(' || line.charAt(i) == '['){
                    parentheses.push(line.charAt(i));
                } else if (line.charAt(i) == ')'){
                    if(parentheses.isEmpty() || parentheses.pop() != '('){
                        checker = false;
                        break;
                    }
                } else if (line.charAt(i) == ']'){
                    if(parentheses.isEmpty() || parentheses.pop() != '['){
                        checker = false;
                        break;
                    }
                }
                i += 1;
            }
            if(!parentheses.isEmpty())
                checker = false;
            bw.write((checker ? "yes" : "no") + "\n");
            line = br.readLine();
        }

        bw.flush();
        br.close();
        bw.close();
    }
}